File: src/main/cpp/SwerveModule.cpp
#include "SwerveModule.h"
#include "frc/smartdashboard/SmartDashboard.h"

SwerveModule::SwerveModule(const double Module[] ):
                                                 m_DriveMotor{ (int)Module[0]},
                                                 m_AngleMotor{ (int)Module[1] },
                                                 m_AngleEncoder{ (int)Module[2] },
                                                 m_AngleOffset{ Module[3] },
                                                 m_Feedforward{SwerveConstants::DriveKS, SwerveConstants::DriveKV, SwerveConstants::DriveKA}
{
    //Config Angle Encoder
    m_AngleEncoder.ConfigFactoryDefault();
    m_AngleEncoder.ConfigAllSettings(m_Settings.SwerveCanCoderConfig);
    m_AngleEncoder.ConfigMagnetOffset(m_AngleOffset.value());
    //Config Angle Motor
    m_AngleMotor.ConfigFactoryDefault();
    m_AngleMotor.ConfigAllSettings(m_Settings.SwerveAngleFXConfig);
    m_AngleMotor.SetInverted(SwerveConstants::AngleMotorInvert);
    m_AngleMotor.SetNeutralMode(ctre::phoenix::motorcontrol::NeutralMode::Coast);
     m_AngleMotor.SetSelectedSensorPosition((DegreesToFalcon(0_deg - GetCANCoder().Degrees()) ));






    //Config Drive Motor
    m_DriveMotor.ConfigFactoryDefault();
    m_DriveMotor.ConfigAllSettings(m_Settings.SwerveDriveFXConfig);
    m_DriveMotor.SetInverted(SwerveConstants::DriveMotorInvert);
    m_DriveMotor.SetNeutralMode(ctre::phoenix::motorcontrol::NeutralMode::Brake);
    m_DriveMotor.SetSelectedSensorPosition(0);
    m_DriveMotor.EnableVoltageCompensation(true);

    m_LastAngle = GetState().angle.Degrees();
    
}



void SwerveModule::SetDegrees(units::degree_t Degrees){

    m_AngleMotor.SetSelectedSensorPosition( (DegreesToFalcon(Degrees - GetCANCoder().Degrees()) ));

}

double SwerveModule::getTurnCounts(){
    return m_AngleMotor.GetSelectedSensorPosition();
}
void SwerveModule::SwapOrientation(){
    m_DriveMotor.SetInverted(!m_DriveMotor.GetInverted());
}

void SwerveModule::SetDesiredState(frc::SwerveModuleState& DesiredState, bool IsOpenLoop){
    DesiredState = Optimize(DesiredState, GetState().angle);
    if(IsOpenLoop){
        double PercentOutput = DesiredState.speed / SwerveConstants::MaxSpeed;
        m_DriveMotor.Set(PercentOutput);
    }else{
        double Velocity = MPSToFalcon(DesiredState.speed);  
        double VoltageFeedForward = m_Feedforward.Calculate(DesiredState.speed)/SwerveConstants::kNominal;
        m_DriveMotor.Set(ctre::phoenix::motorcontrol::ControlMode::Velocity, Velocity, ctre::phoenix::motorcontrol::DemandType_ArbitraryFeedForward, VoltageFeedForward);
    }
    units::meters_per_second_t minSpeed = (SwerveConstants::MaxSpeed * 0.01);

    units::degree_t Angle = units::math::abs(DesiredState.speed) <= minSpeed ? m_LastAngle: DesiredState.angle.Degrees();

    m_AngleMotor.Set( ctre::phoenix::motorcontrol::ControlMode::Position, DegreesToFalcon(Angle) );
    m_LastAngle = Angle;
}

units::degree_t SwerveModule::getLastAngle(){
    return m_LastAngle;
}

/* This custom optimize method is created because Wpilib assumes the controller is continuous, which the CTRE Talons are not. */
frc::SwerveModuleState SwerveModule::Optimize(frc::SwerveModuleState DesiredState, frc::Rotation2d CurrentAngle){


    units::degree_t ModReferenceAngle { frc::AngleModulus( CurrentAngle.Radians() )  };
    frc::SmartDashboard::SmartDashboard::PutNumber("Current Angle", ModReferenceAngle.value());
    frc::SmartDashboard::SmartDashboard::PutNumber("Desired Angle(Continuous)", DesiredState.angle.Degrees().value());

    units::meters_per_second_t TargetSpeed = DesiredState.speed;
    units::degree_t Delta = DesiredState.angle.Degrees() - ModReferenceAngle;
    if(Delta >= 270_deg){
        Delta -= 360_deg;
    }else if(Delta <= -270_deg){
        Delta += 360_deg;
    }
    if( units::math::abs(Delta) > 90_deg){
        TargetSpeed = - TargetSpeed;
        Delta = Delta > 0_deg ? (Delta -= 180_deg) : ( Delta += 180_deg);
    }
    units::degree_t TargetAngle = CurrentAngle.Degrees() + Delta;
    frc::SmartDashboard::SmartDashboard::PutNumber("Desired Angle(Discontinuous)", DesiredState.angle.Degrees().value());

    return  {TargetSpeed, TargetAngle};
}


frc::Rotation2d SwerveModule::GetCANCoder(){
    return frc::Rotation2d( units::degree_t( m_AngleEncoder.GetAbsolutePosition() ) );
}

frc::SwerveModulePosition SwerveModule::GetPosition(){
    units::meter_t Distance{ FalconToMeters(m_DriveMotor.GetSelectedSensorPosition())};
    frc::Rotation2d Angle{FalconToDegrees( m_AngleMotor.GetSelectedSensorPosition()) };
    return {Distance, Angle};

}

units::meter_t SwerveModule::FalconToMeters(double Counts){
    return units::meter_t{ (Counts * SwerveConstants::WheelCircumference) / ( 2048 * SwerveConstants::DriveGearRatio)};
}

frc::SwerveModuleState SwerveModule::GetState(){
    units::meters_per_second_t Velocity{ FalconToMPS(m_DriveMotor.GetSelectedSensorVelocity(0)) };
    frc::Rotation2d Angle{FalconToDegrees( m_AngleMotor.GetSelectedSensorPosition()) };
    return {Velocity, Angle};
}

units::degree_t SwerveModule::FalconToDegrees(double Counts){
    return units::degree_t(Counts * ( 360.0 / (SwerveConstants::AngleGearRatio * 2048.0)));
}

double SwerveModule::DegreesToFalcon(units::degree_t Degrees){
    return Degrees.value() / (360.0 / (SwerveConstants::AngleGearRatio * 2048.0));
}

double SwerveModule::FalconToRPM(double VelocityCounts){
    double MotorRPM = VelocityCounts * ( 600.0 / 2048.0);
    double MechRPM = MotorRPM / SwerveConstants::DriveGearRatio;
    return MechRPM;
}

double SwerveModule::RPMToFalcon(double RPM){
    double MotorRPM = RPM * SwerveConstants::DriveGearRatio;
    double SensorCounts = MotorRPM * (2048.0 / 600.0);
    return SensorCounts;
}

units::meters_per_second_t SwerveModule::FalconToMPS(double VelocityCounts){
    double WheelRPM = FalconToRPM(VelocityCounts);
    units::meters_per_second_t WheelMPS = (WheelRPM * SwerveConstants::WheelCircumference) /60_s;
    return WheelMPS;
}

double SwerveModule::MPSToFalcon(units::meters_per_second_t Velocity){
    double WheelRPM = ( Velocity.value() * 60 ) / SwerveConstants::WheelCircumference.value();
    double WheelVelocity = RPMToFalcon(WheelRPM);
    return WheelVelocity;
}

File: src/main/java/frc/robot/SwerveModule.java

package frc.robot;
import com.ctre.phoenix.motorcontrol.ControlMode;
import com.ctre.phoenix.motorcontrol.DemandType;
import com.ctre.phoenix.motorcontrol.NeutralMode;
import com.ctre.phoenix.motorcontrol.can.WPI_TalonFX;
import com.ctre.phoenix.sensors.WPI_CANCoder;
import edu.wpi.first.math.controller.SimpleMotorFeedforward;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.kinematics.SwerveModuleState;
import edu.wpi.first.math.kinematics.SwerveModulePosition;
import edu.wpi.first.math.MathUtil;
// import frc.robot.HardwareConfig;
import frc.robot.Constants.SwerveConstants;


public class SwerveModule {
    private WPI_TalonFX m_DriveMotor;
    private WPI_TalonFX m_AngleMotor;
    private WPI_CANCoder m_CANCoder;
    private double m_LastAngle;
    private double m_AngleOffset;
    private SimpleMotorFeedforward m_Feedforward;
    private HardwareConfig m_Settings;

    public SwerveModule(double[] module){
        m_DriveMotor = new WPI_TalonFX((int)module[0], "DriveCANivore");
        m_AngleMotor = new WPI_TalonFX((int)module[1], "DriveCANivore");
        m_CANCoder = new WPI_CANCoder((int)module[2], "DriveCANivore");
        m_AngleOffset = module[3];
        m_Feedforward = new SimpleMotorFeedforward(SwerveConstants.driveKS, SwerveConstants.driveKV, SwerveConstants.driveKA);
        m_Settings = new HardwareConfig();

        /* CANCoder Config */
        m_CANCoder.configFactoryDefault();
        m_CANCoder.configAllSettings(m_Settings.swerveCANCoderConfig);

        /* Angle Motor Config */
        m_AngleMotor.configFactoryDefault();
        m_AngleMotor.configAllSettings(m_Settings.swerveAngleFXConfig);
        m_AngleMotor.setInverted(SwerveConstants.angleMotorInvert);
        m_AngleMotor.setNeutralMode(NeutralMode.Coast);
        m_AngleMotor.setSelectedSensorPosition(degreesToFalcon(getCANCoder().getDegrees() - m_AngleOffset));

        /* Drive Motor Config */
        m_DriveMotor.configFactoryDefault();
        m_DriveMotor.configAllSettings(m_Settings.swerveDriveFXConfig);
        m_DriveMotor.setInverted(SwerveConstants.driveMotorInvert);
        m_DriveMotor.setNeutralMode(NeutralMode.Brake);
        m_DriveMotor.setSelectedSensorPosition(0.0);
        m_DriveMotor.enableVoltageCompensation(true);

        m_LastAngle = getState().angle.getDegrees();
    }

    public double getTurnCounts(){
        return m_AngleMotor.getSelectedSensorPosition();
    }

    public void setDesiredState(SwerveModuleState desiredState, boolean isOpenLoop){
        desiredState = optimize(desiredState, getState().angle);
        if(isOpenLoop){
            double percentOutput = desiredState.speedMetersPerSecond / SwerveConstants.maxSpeed;
            m_DriveMotor.set(percentOutput);
        } else {
            double velocity = mpsToFalcon(desiredState.speedMetersPerSecond);
            double voltageFeedForward = m_Feedforward.calculate(desiredState.speedMetersPerSecond) / SwerveConstants.kNominal;
            m_DriveMotor.set(ControlMode.Velocity, velocity, DemandType.ArbitraryFeedForward, voltageFeedForward);
        }
        double minSpeed = SwerveConstants.maxSpeed * 0.01;
        double angle = Math.abs(desiredState.speedMetersPerSecond) <= minSpeed ? m_LastAngle : desiredState.angle.getDegrees();
        m_AngleMotor.set(ControlMode.Position, degreesToFalcon(angle));
        m_LastAngle = angle;
    }

    public void setDesiredAngle(Rotation2d angle){
        SwerveModuleState tempState = new SwerveModuleState(0.0, angle);
        tempState = optimize(tempState, getState().angle);
        m_AngleMotor.set(ControlMode.Position, degreesToFalcon(angle.getDegrees()));
        m_LastAngle = angle.getDegrees();
    }

    public double getLastAngle(){
        return m_LastAngle;
    }

    public SwerveModuleState optimize(SwerveModuleState desiredState, Rotation2d currentAngle){
        double modReferenceAngle = MathUtil.angleModulus(currentAngle.getRadians()) * 180.0 / Math.PI;
        double targetSpeed = desiredState.speedMetersPerSecond;
        double delta = desiredState.angle.getDegrees() - modReferenceAngle;
        if(delta >= 270.0){
            delta -= 360.0;
        }
        if(delta <= -270.0){
            delta += 360.0;
        }
        if(Math.abs(delta) > 90.0){
            targetSpeed *= -1.0;
            delta = delta > 0.0 ? delta - 180.0 : delta + 180.0;
        }
        double targetAngle = currentAngle.getDegrees() + delta;

        return new SwerveModuleState(targetSpeed, Rotation2d.fromDegrees(targetAngle));
    }

    public void resetToAbsolute(){
        m_AngleMotor.setSelectedSensorPosition(degreesToFalcon(getCANCoder().getDegrees() - m_AngleOffset));
    }

    public Rotation2d getCANCoder(){
        return Rotation2d.fromDegrees(m_CANCoder.getAbsolutePosition());
    }

    public SwerveModulePosition getPosition(){
        double distance = falconToMeters(m_DriveMotor.getSelectedSensorPosition());
        Rotation2d angle = Rotation2d.fromDegrees(falconToDegrees(m_AngleMotor.getSelectedSensorPosition()));
        return new SwerveModulePosition(distance, angle);
    }

    public SwerveModuleState getState(){
        double velocity = falconToMPS(m_DriveMotor.getSelectedSensorVelocity(0));
        Rotation2d angle = Rotation2d.fromDegrees(falconToDegrees(m_AngleMotor.getSelectedSensorPosition()));
        return new SwerveModuleState(velocity, angle);
    }

    public double falconToMeters(double counts){
        return (counts * SwerveConstants.wheelCircumference) / (2048.0 * SwerveConstants.driveGearRatio);
    }

    public double falconToDegrees(double counts){
        return counts * (360.0 / (SwerveConstants.angleGearRatio * 2048.0));
    }

    public double degreesToFalcon(double degrees){
        return degrees / (360.0 / (SwerveConstants.angleGearRatio * 2048.0));
    }

    public double falconToRPM(double velocityCounts){
        double motorRPM = velocityCounts * (600.0 / 2048.0);
        return motorRPM / SwerveConstants.driveGearRatio;
    }

    public double rpmToFalcon(double rpm){
        double motorRPM = rpm * SwerveConstants.driveGearRatio;
        return motorRPM * (2048.0 / 600.0);
    }

    public double falconToMPS(double velocityCounts){
        double wheelRPM = falconToRPM(velocityCounts);
        return wheelRPM * SwerveConstants.wheelCircumference / 60.0;
    }

    public double mpsToFalcon(double mps){
        double wheelRPM = (mps * 60.0) / SwerveConstants.wheelCircumference;
        return rpmToFalcon(wheelRPM);
    }
}
